### PR TITLE
Fix to infinite loading on approve screen

### PR DESCRIPTION
**Pull request by adonesky1**
Fixes issue where screen for confirming `approve()` method calls gets stuck in infinite loading (https://github.com/MetaMask/metamask-extension/issues/14459 & https://github.com/MetaMask/metamask-extension/issues/14589 & https://github.com/MetaMask/metamask-extension/issues/14518). This is set to be partially addressed when the commit associated with this [PR](https://github.com/MetaMask/controllers/pull/834) is merged into the extension. It will be further addressed something like this [WIP PR](https://github.com/MetaMask/controllers/pull/838) (the exact form of which is still being discussed). Even with these fixes however, there are still contracts for which the issue will not be resolved. The ENS ERC721 contract for example does not implement either a `symbol()` or `name()` method the result of which the approve screen currently relies upon. This PR simply adds a fallback value of `""` for the symbol so that once we're done loading contract data the symbol will not be undefined no matter what, [allowing us to move past the loading screen.](https://github.com/MetaMask/metamask-extension/blob/634cf70a71f63ec98ff7efbc081620480efa8a58/ui/pages/confirm-approve/confirm-approve.js#L152)
 
This fixes a regression introduced in this PR: https://github.com/MetaMask/metamask-extension/pull/13788

### CODE REVIEW

**Review by darkwing — DISMISSED** [review]

**Comment by adonesky1** [comment]
While this is definitely better than infinite loading we should probably do a bit of UX improvement for the cases where someone is calling `approve` on a token contract that does not match one of the three supported contract types (ERC20, ERC721, ERC1155)

Screen Shot 2022-05-23 at 11 17 27 AM

I know we're going to redesign these screens pretty soon so probably not worth a ton of effort but something small to help us cross the bridge for now? Maybe just a copy change saying something like: "Unable to determine token standard, name or symbol. Approving will grant allowance for this unidentifiable token to the following address." ?

@rachelcope @holantonela

**Comment by rachelcope** [comment]
How about `Give permission to access your token`?

**Review by ryanml — DISMISSED** [review]

**Comment by adonesky1** [comment]
Updated this fix to include additional context for the user to be sure they know what they are approving in cases where we don't have complete information:

[Dentraland (LAND)](https://etherscan.io/address/0xf87e31492faf9a91b02ee0deaad50d51d56d5d4d#writeProxyContract):
Doesn't fully implement the metadata interface, will be further fixed when https://github.com/MetaMask/controllers/pull/834 is pulled into the extension:
https://user-images.githubusercontent.com/34557516/170384627-d8df6e74-61bd-4cdd-b6c1-75378c5aad18.mov

**Comment by darkwing** [comment]
Code looks good and works, but I"m wondering if there's a test we can add to prevent regressions, since the ERC logic is and symbol asset stuff presents many cases.

**Comment by adonesky1** [comment]
@darkwing I believe @SayaGT has a new design for the approve screen in the works. I'm happy to add tests, just have been a bit hesitant to pour too much effort into this since we're going to rebuild it pretty soon as I understand it.

**Comment by binzydev** [comment]
Pour please 🍷 🙏  

More than a few projects halted due to this issue and even if you rebuild it, I assume you'll want a stable/LTS legacy just incase because with changes like this it's breaking support for who knows how many projects.

**Comment by adonesky1** [comment]
> I think a test would be helpful -- do we know when that redesign is scheduled for? Even still, I think it could take months for that new UI to hit.

Struggling to put together a useful test for this... It seems really testing this functionality involves testing `confirm-token-transaction-switch.js` and routing correctly in a test is proving absurdly difficult. Been hitting my head against the wall for a day now. Any thoughts here @darkwing?

**Review by darkwing — DISMISSED** [review]
I was hoping to find a test for `getTitleTokenDescription` since there are so many iterations.  Maybe we could export this to a helper function in the future to make testing easier?  All my testing worked though.

**Comment by adonesky1** [comment]
> I was hoping to find a test for getTitleTokenDescription since there are so many iterations.

@darkwing thanks for your input, I will add a test for `getTitleTokenDescription`

**Comment by adonesky1** [comment]
I think additionally we should add e2e tests for this flow.

**Comment by adonesky1** [comment]
@darkwing I actually don't think we should extract `getTitleTokenDescription` as a helper function. Won't realistically be used anywhere else. I'm turning my efforts to adding a non-standard token flow in the test-dapp instead, which will add coverage for this functionality much better anyways

**Comment by adonesky1** [comment]
@darkwing added test coverage to partially cover us against regressions for this: [useAssetDetails test](https://github.com/MetaMask/metamask-extension/pull/14756/commits/025f32a5cb5915eb8f8c12bc97dfa26b47b679e3)

but I am also working on adding non-standard tokens e2e tests, with some additions to test-dapp

**Comment by mcmire** [comment]
@adonesky1 It looks like you're `await`ing `renderUseAssetDetails` in your test. I think if you take this off it should allow you to `await waitForNextUpdate()` right after that.

**Comment by adonesky1** [comment]
> It looks like you're awaiting renderUseAssetDetails in your test. I think if you take this off it should allow you to await waitForNextUpdate() right after that

Unfortunately I'm still getting the warning (`Warning: An update to TestHook inside a test was not wrapped in act(...)`) when using this pattern

**Comment by mcmire** [comment]
@adonesky1 `waitForNextUpdate` is a function. So you need to say `await waitForNextUpdate()`.

**Comment by mcmire** [comment]
I changed the first test to this and it seemed to work:

``` javascript
  it('should return object with tokenSymbol set to and empty string, when getAssetDetails returns and empty object', async () => {
    const toAddress = '000000000000000000000000000000000000dead';
    const tokenAddress = '0x1';

    const transactionData = `0xa9059cbb000000000000000000000000${toAddress}000000000000000000000000000000000000000000000000016345785d8a0000`;

    const { result, waitForNextUpdate } = renderUseAssetDetails({
      tokenAddress,
      userAddress: '0x111',
      transactionData,
    });

    await waitForNextUpdate();

    expect(result.current).toStrictEqual({
      assetAddress: tokenAddress,
      assetName: undefined,
      assetStandard: undefined,
      decimals: undefined,
      toAddress: `0x${toAddress}`,
      tokenAmount: undefined,
      tokenId: undefined,
      tokenImage: undefined,
      tokenSymbol: '',
      tokenValue: undefined,
      userBalance: undefined,
    });
  });
```

**Comment by adonesky1** [comment]
> @adonesky1 waitForNextUpdate is a function. So you need to say await waitForNextUpdate().

😜 I meant to do that... whoops! Thanks for the call out!

**Comment by adonesky1** [comment]
> @adonesky1 waitForNextUpdate is a function. So you need to say await waitForNextUpdate().

@mcmire fixed here: https://github.com/MetaMask/metamask-extension/pull/14756/commits/d8e17ef674fde526bba88c5b7e1d12d5994b6d6c

**Comment by mcmire** [comment]
@adonesky1 Nice! That matches what I was able to do locally :)

File: ui/hooks/useAssetDetails.js [hunk 1]
```diff
@@ -1,24 +1,21 @@
...
-import { getTokenList } from '../selectors';
 import { hideLoadingIndication, showLoadingIndication } from '../store/actions';
 import { usePrevious } from './usePrevious';
 
 export function useAssetDetails(tokenAddress, userAddress, transactionData) {
   const dispatch = useDispatch();
   // state selectors
-  const tokens = useSelector(getTokens);
   const collectibles = useSelector(getCollectibles);
-  const tokenList = useSelector(getTokenList);
```
Comment (adonesky1) [inline ui/hooks/useAssetDetails.js:21]: `tokens` and `tokenList` were never used by this hook, they had been at one point while developing it initially, but can be safely removed.

**Review by darkwing — APPROVED** [review]

**Review by danjm — APPROVED** [review]
LGTM, nice work

**Pull request merged**